enhancements/windows-containers/container-runtime-containerd.md

```diff
@@ -0,0 +1,205 @@
+Current CNI/IPAM will be used for containerd and no change in HNS-Network and HNS-Endpoint creation
+steps. The config file which will be used by containerd will point to the same CNI/IPAM executables.
+
+## WMCO CLI option
+WMCO CLI option 'runtime' will be used to make containerd as the default runtime.  In the 4.10 community operator, it
+will be specified as `--runtime containerd`. In the absence of this flag, Docker will become the default runtime. We remove
+this flag in 4.11 and make containerd the default runtime. At any given point in time, we have one runtime support and
+do not allow switching between runtime.
+
+## Containerd logging
+Containerd can be started with parameters in which we can enable logging and specify the file path to
+log warnings/errors. Log files will be stored at c:\var\log\containerd. Pod logs will be stored in the
+same location (C:\var\log\pods)
+
+## Upgrade
+WMCO supports upgrade. Please refer [WMCO-Upgrade](https://github.com/openshift/enhancements/blob/master/enhancements/windows-containers/windows-machine-config-operator-upgrades.md)
+enhancement proposal document to learn more about WMCO upgrade.
+
+The procedure for an upgrade is as follows:
+1) MachineSet
+   * As part of upgrade Machine object gets deleted, resulting in the drain and deletion of the Windows node.
+   * New Windows VM will be created.
+   * The upgraded WMCO instance will configure Windows VM with containerd as the default runtime.
+   * Once the Windows node joins the Kubernetes cluster, the user can deploy pods on the Windows node.
+2) BYOH ( Bring Your Own Host)
+   * upgrade process will uninstall kubelet, kube-proxy, CNI, and hybrid-overlay components that were installed by WMCO.
+   * Windows OS-specific configurations like HNS-Network that were created by WMCO instance would be deleted.
+   * Once cleanup is done, WMCO instance will install containerd along with kubelet, kube-proxy, CNI and
+     hybrid-overlay.
+   * Kuebelet will start as a service. Containerd will be the default runtime.
+   * Once the Windows node joins the Kubernetes cluster, the user can deploy pods on the Windows node.
+
+### Risks and Mitigations
+
+* If the cluster is upgraded, and the new version introduces an issue due to containerd
+  that should be addressed because Docker runtime support won't be available in kubelet.
+  As we don't support downgrade, reverting to an older version is not possible. To overcome
+  any issues, we introduce this feature as part of the 5.0 community operator so that this feature
+  can be well tested before we make it default in WMCO 6.0 release. We will use two test jobs to test all e2e test cases
+  in the 4.10 community branch(OKD) and 4.10 branch(ccm OCP)to capture any regression.
+  This will help us to identify any issues well before containerd becomes default runtime.
+* If we run into any issue in containerd, we work closely with the containerd open source community and try
+  to resolve it as soon as possible.
```

Comment:
What is our patching strategy for containerd? Are we going to be using a carry patch strategy, where we are only downstreaming changes that we open upstream PRs for? What happens if we need specific functionality within containerd that upstream does not agree with? Will we end up diverging from upstream?

Reply:
Containerd is a submodule and the same rule of how we handle other submodules will be applicable to contained as well. 
In short, we will be updating submodules once in every sprint. If there is an emergency fix that comes in, we shall update the submodule on a need basis. We will not fix anything in donwstream. if we fix any issue, it should be fixed first in upstream and we should rebase downstream.

Reply:
@selansen, @sebsoto is asking about carry patches i.e. say a fix is not accepted in time upstream but we need to pull it sooner. We can pull the fix into the downstream containerd repo given that is where we point the submodule at. I am not saying we should do this but we need to call out what our strategy here is. 

>In short, we will be updating submodules once in every sprint. 

We aim to update submodules on a weekly basis for the main K8s components. For containerd we need to come up with a cadence given we own the downstream repo and need to pull from upstream.

Reply:
Containerd is going to be our runtime for a long time. So it is better to work closely with the community and push the fix in the upstream first and then bring it down to downstream. if we have any critical issue in contained, other vendors who use contained also will face the same issue. It will be good to get community support for any critical issue.
I will update this in the document. 
@aravindhp @sebsoto if you have a different opinion, pls do let me know.